plugins/experimental/url_sig/url_sig.c

```diff
@@ -330,23 +330,50 @@ getAppQueryString(const char *query_string, int query_length)
   }
 }
 
+/** fixedBufferWrite safely writes no more than *dest_len bytes to *dest_end
+ * from src. If copying src_len bytes to *dest_len would overflow, it returns
+ * zero. *dest_end is advanced and *dest_len is decremented to account for the
+ * written data. No null-terminators are written automatically (though they
+ * could be copied with data).
+ */
+static int
+fixedBufferWrite(char **dest_end, int *dest_len, const char *src, int src_len)
+{
+  if (src_len > *dest_len) {
+    return 0;
+  }
+  memcpy(*dest_end, src, src_len);
+  *dest_end += src_len;
+  *dest_len -= src_len;
+  return 1;
+}
+
 static char *
 urlParse(char *url, char *anchor, char *new_path_seg, int new_path_seg_len, char *signed_seg, unsigned int signed_seg_len)
 {
   char *segment[MAX_SEGMENTS];
   unsigned char decoded_string[2048] = {'\0'};
-  char new_url[8192]                 = {'\0'};
+  char new_url[8192]; /* new_url is not null_terminated */
   char *p = NULL, *sig_anchor = NULL, *saveptr = NULL;
-  int i = 0, numtoks = 0, cp_len = 0, l, decoded_len = 0, sig_anchor_seg = 0;
+  int i = 0, numtoks = 0, decoded_len = 0, sig_anchor_seg = 0;
+
```

Comment:
could use nullchecks on all the things

Reply:
No null checks in the old things, and my goal wasn't wholesale rewrite, but just a return to existing behavior. These are pretty safe, though, cause if we get a null ptr here, we're already in pretty bad shape much earlier.